### ISSUE TITLE
Max instructions reached error while running lua script in Ethos 1.5.15 

### ISSUE THREAD

**Issue by fdm225** [FrSkyRC/ETHOS-Feedback-Community#4505]
Hi Devs,
I tried to update issue 4103 but didn't see any official response and would like to try and get this resolved.  I am fixing a script that generates a QR code from the last logged GPS location and it does a lot of processing when started, enough to trigger the Max instructions count reached message. **I am doing the processing in the wakeup function on version 1.5.15**; do I need to do anything to use the preemption that is discussed in issue 4103 or am I misunderstanding how it works? I assumed that when the instruction count was reached it would preempt the current running script and when wakeup is called again it would pickup from where it left off, is that correct?  I have attached the modified script in the case that it is of use.

```
local function wakeup(widget)
	doHeavyCalc(widget)
```

![image](https://github.com/user-attachments/assets/64a13a17-9e11-4854-8368-fde9b7c6a78a)
[Ethos-ModelFinder.zip](https://github.com/user-attachments/files/16947471/Ethos-ModelFinder.zip)

**Comment by strgaltdel** [comment]
just some additional info

- the QR code calculation is based on "speedata's" open source coding from here: https://github.com/speedata/luaqrcode/blob/master/qrencode.lua 

- can confirm, QR code procurement in wakeup handler runs into "max instructions" error

- this script is 2 years and needs some more tweaking then update QR calculation,. Due to lack of knowledge at this time, there are some "don't do this" implemented, unfortunately too much projects in parallel......



this is a more condensed script, focussing on the max instructions issue
just calculating the QR code (in wakeup handler) in case SF was switched

[main.zip](https://github.com/user-attachments/files/16959793/main.zip)


regards

**Comment by fdm225** [comment]
I fixed a couple of small issues but when I ran into the instruction limit.  I attempted to move it into the wakeup function and am putting any additional fixes on hold until I see if there is a resolution.  If the only fix is to rewrite the computation portion of the QR code I will probably let this go as I do not want to take the time to understand where to break the main computation code up.

dave

**Comment by bsongis-frsky** [comment]
You have attached several scripts, I see one source and one widget. 
Is it possible that you attach the minimum to reproduce the issue?

**Comment by strgaltdel** [comment]
... you can take my "main.zip"-widget "QR-Calc" from above (only one main.lua), 
it's a subset of fdm225's widget, my example does the QR calculation only as a "proof of concept" and triggers the failure in wakeup handler too

Calculation (and failure) is initiated by switching SF in Sim.

![image](https://github.com/user-attachments/assets/32373db2-6cf3-4088-8933-ab0b907f6102)

regards

**Comment by strgaltdel** [comment]
just drilled it a little bit down:

after a lot of precalculation, the QR code calc stucks in the "add_data_to_matrix" function.
he performed app. 192 loops in this function until the "too much instructions" trigger steps in.

but again, at this point a lot of precalc was done before calling this function ....

regards
udo


```lua
-- Add the data string (0's and 1's) to the matrix for the given mask.
local function add_data_to_matrix(matrix,data,mask)
	size = #matrix
	local x,y,positions
	local debugLoop = 1
	local _x,_y,m
	local dir = "up"
	local byte_number = 0
	x,y = size,size
	string.gsub(data,".?.?.?.?.?.?.?.?",function ( byte )
		byte_number = byte_number + 1
		positions,x,y,dir = get_next_free_positions(matrix,x,y,dir,byte,mask)
		for i=1,#byte do
			_x = positions[i][1]
			_y = positions[i][2]
			print("QR: add data string 2 matrix, run:",debugLoop)
			debugLoop = debugLoop+1
			m = get_pixel_with_mask(mask,_x,_y,string.sub(byte,i,i))
			if debugging then
				matrix[_x][_y] = m * (i + 10)
			else
				matrix[_x][_y] = m
			end
		end
	end)
end
```
